core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,226 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Closeable;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ *
+ *
+ */
+// Discussion point. Using java.io.Closeable here so we can inherit AutoCloseable where its available.
+public final class Context implements Closeable {
+
+  /**
+   * The logical root context which is {@link #current()} if no other context is bound. This context
+   * is already closed and will not retain any listeners.
+   */
+  public static final Context ROOT = new Context();
+  static {
+    ROOT.close();
+  }
+
+  /**
+   * A listener notified on context {@link #close()}.
+   */
+  public interface Listener {
+    /**
+     * @param context the newly closed context.
+     */
+    public void closed(Context context);
+  }
+
+  /**
+   * Create a new context with the given key value set.
+   */
+  public <V> Context withValue(Key<V> k1, V v1) {
+    throw new UnsupportedOperationException();
+  }
+
+  public <V1, V2> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2) {
+    throw new UnsupportedOperationException();
+  }
+
+  public <V1, V2, V3> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2, Key<V3> k3, V3 v3) {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Create a new context with the given listener
+   */
+  public <V> Context withListener(Listener listener) {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Create a new context that does not cascade closure from its parent
+   */
+  public <V> Context withNoCascade() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Create a new context that cascades closure from its parent
+   */
+  public <V> Context withCascade() {
```

Comment:
Cascading will apply by default (will clarify that in the javadoc and will remove withCascade which is really just an artifact for testing).

... and yes ordering is very important, will add to doc as well.